### ISSUE TITLE
`unnecessary-list-index-lookup` false-positive when iterable is modified in loop

### ISSUE THREAD

**Issue by The-Compiler** [PyCQA/pylint#6896]
### Bug description

With some code like this:

```python
# pylint: disable=missing-module-docstring

parts = ["a", "b", "c", "d"]
for i, part in enumerate(parts):
    if i == 3:  # more complex condition actually
        parts.insert(i, "X")
    print(part, parts[i])
```

([real-world example](https://github.com/qutebrowser/qutebrowser/blob/1996c4d7e72bb3764cd4ea316f8cf84d4fc32dca/qutebrowser/completion/completer.py#L154-L165)), pylint claims that `parts[i]` could be replaced by `part` - but as running the file shows, that isn't actually the case:

```
a a
b b
c c
d X
d d
```

You might think that this is a bit unorthodox (and I might agree, should probably rewrite this...), but if I'm reading the [Python docs](https://docs.python.org/3.7/reference/compound_stmts.html#the-for-statement) right, the behavior is well-defined (emphasis mine):

> There is a subtlety when the sequence is being modified by the loop (this can only occur for mutable sequences, e.g. lists). An internal counter is used to keep track of which item is used next, and this is incremented on each iteration. When this counter has reached the length of the sequence the loop terminates. This means that if the suite deletes the current (or a previous) item from the sequence, the next item will be skipped (since it gets the index of the current item which has already been treated). Likewise, **if the suite inserts an item in the sequence before the current item, the current item will be treated again the next time through the loop**.

cc @timmartin who added the checker in #5834.

### Command used

```shell
pylint x.py
```

### Pylint output

```shell
************* Module x
x.py:7:16: R1736: Unnecessary list index lookup, use 'part' instead (unnecessary-list-index-lookup)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.14.1
astroid 2.11.5
Python 3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0]
```

### OS / Environment

Archlinux

**Comment by The-Compiler** [comment]
Also see #6818, though this seems subtly different. Note that #6845 doesn't seem to fix this.

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue and providing the relevant doc I think this is indeed a false positive (even if I personally would not modify iterable while iterating them to make the code as unsurprising as possible).
